### PR TITLE
[SPMD] Output param sharding

### PR DESCRIPTION
**Pull request by yeounoh**
This is a follow-up work of #4721 , to shard the computation outputs. This will help us avoid resharding the replicated outputs for more efficient stepping. Most notably, 
- we introduce `ShardingUtil::PrepareOutputShardingPropagation` to extract `SharidngSpecPtr` from sync tensors and prepare `PjRtShardedData` placeholders. 
- remove the logic to prepare PjRtShardedData placeholders in `ExtractIRAndPrepareXlaData_` 
- remove `CollectSharidngSpecs` also in favor of `PrepareOutputShardingPropagation`.

### CODE REVIEW

File: third_party/xla_client/pjrt_computation_client.cc
Comment (JackCaoG) [inline third_party/xla_client/pjrt_computation_client.cc:352]: maybe you should update the comment here as well.

File: torch_xla/csrc/ops/device_data.h
Comment (JackCaoG) [inline torch_xla/csrc/ops/device_data.h:30]: we should at very least check if shape matches before modifying the underlying data

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:1016]: shouldn't we check whether graph is sharded before calling these functions?

Comment (yeounoh) [reply torch_xla/csrc/xla_graph_executor.cpp:1016]: This call is a no-op if the graph is not sharded, as this is used in the SyncTensors code path.

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (jonb377) [inline torch_xla/csrc/xla_sharding_util.cpp:413]: In `xla_graph_executor.cpp`, the sharded data preparation for implicitly replicated data is removed. Should we still prepare a sharded placeholder regardless of `tuple_shardings[i].type()`?

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (JackCaoG) [inline torch_xla/csrc/xla_sharding_util.cpp:430]: in what case output tensor will not be sharded? I thought under SPMD it is either sharded or replicated.

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:555]: any chance we can undo these comment changes? They make reviewing a bit difficult

File: torch_xla/csrc/tensor.cpp
Comment (JackCaoG) [inline torch_xla/csrc/tensor.cpp:454]: maybe undo this

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:413]: In summary, we shouldn't prepare the sharded data placeholder in the graph executor, but do so using this `ShardingUtil` API. This happens based on `tuple_shardings[i].type`, not "regardless of" it. Um, let me know if you still have questions.

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (JackCaoG) [reply torch_xla/csrc/xla_graph_executor.cpp:1016]: right, I would still prefer put this under a if condition, because
1. it is easy to read, I don't need to go into `PrepareOutputShardingPropagation` and figured that it is a no-op if it is not sharded
2. we don't need to allocate `sharding_specs` for non-SPMD
3. if one day we update `PrepareOutputShardingPropagation` and forgot about the caller always call(even for non-spmd) we might introduce weird issue.

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:430]: Correct. When it's replicated, it's no longer sharded. If user decides to unshard or the compiler decides to remove the propagation, then we that needs to be reflected on the XLATensor's as source of the truth. This is a safeguard for the edge use cases.

Comment (JackCaoG) [reply torch_xla/csrc/xla_sharding_util.cpp:430]: oh ok, when you say it is not `sharded`, I thought it means it does not has sharding, hence it won't even be replicated. I guess comment here can be more specified about the replication.

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (jonb377) [reply torch_xla/csrc/xla_sharding_util.cpp:413]: Just for my understanding - in the implicit replication case, the data is wrapped in `PjRtShardedData` and `xtensor->sharding_spec()` remains nullptr. In the `OutputHandler`, we're wrapping the resulting shards in `PjRtShardedData`. If we don't override the placeholder with `PjRtShardedData` here, we'll be trying to `Assign` a `PjRtShardedData` to a `PjRtData` placeholder.

If this is working, it seems like the `tuple_shardings[i].type()` is non-false even for the implicitly replicated case, unless I'm missing something?

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_graph_executor.cpp:1016]: I see, good point. I think a better place would be to initialize and call `PrepareOutputShardingPropagation` if (async->is_sharded...)` just before the execution. Let me make some changes.

Comment (JackCaoG) [reply torch_xla/csrc/xla_graph_executor.cpp:1016]: make sure do that before the `barrier` call in https://github.com/pytorch/xla/blob/master/torch_xla/csrc/xla_graph_executor.cpp#L746, otherwise this will be blocked by the execution previous execution and can not be overlapped (which will increase the overhead of tracing in general).

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (alanwaketan) [inline torch_xla/csrc/xla_sharding_util.cpp:389]: why pointer instead of reference?

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (alanwaketan) [inline torch_xla/csrc/xla_sharding_util.cpp:403]: Log?

**Review by alanwaketan — COMMENTED** [review]
Can you briefly describe the code being removed?

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:389]: No particular or a strong reason, just wanted to be consistent with other inputs `std::vector<XLATensorPtr>* tensors` .

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:403]: The PR's in draft -- I am testing a few extra bits with probing. Will ping you when it's ready :) thanks.

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp]: After merging with the recent implicit replication changes, all the outputs in case of sharded computation should be wrapped inside PjRtShardedData. This is a bug, will update the PR.

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:419]: Great, I am also adding the implicitly replicated PjRtShardedData here after rebasing.

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_graph_executor.cpp]: yea, also sourcing the inputs to the function seems untidy.. going with your original suggestion:
```
// Extract sharding specs for the results and prepare the sharded data
  // placeholders if the computation is sharded.
  if (cached_computation->is_sharded) {
    ShardingUtil::PrepareOutputShardingPropagation(
        tensors, coll->indices, cached_computation->computation, &tensors_data,
        &sharding_specs);
  }
```
we still need to initialize the sharding_specs with null, in case of a unsharded compuation.

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp]: Ok removed the `std::cout`, I will mark this PR as ready-for-review tmr morning.

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_graph_executor.cpp]: Removed unnecessary linting changes.

File: third_party/xla_client/pjrt_computation_client.cc
Comment (yeounoh) [reply third_party/xla_client/pjrt_computation_client.cc]: > Can you briefly describe the code being removed?

Updated the PR description, thanks :)

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (jonb377) [inline torch_xla/csrc/xla_sharding_util.cpp:403]: Could you add a comment explaining what this is testing?

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (jonb377) [inline torch_xla/csrc/xla_graph_executor.cpp:1002]: It seems like we have to override some of the work done in `SetTensorData` in `PrepareOutputShardingPropagation`, is there a reason we can't call `PrepareOutputShardingPropagation` first?

**Review by jonb377 — APPROVED** [review]
LGTM, thanks Yeounoh! This is great

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (alanwaketan) [reply torch_xla/csrc/xla_sharding_util.cpp:389]: We can clean up this later. It seems weird to use raw pointers...

**Review by alanwaketan — APPROVED** [review]
LGTM.

File: torch_xla/csrc/xla_graph_executor.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_graph_executor.cpp:1002]: Good question, and the answer is no (though we can make changes to both `SetTensorData` and `PrepareOutputShardingPropagation` to make it work in the reverse order, more work) sine we still want to hit `handle == nullptr` case there. I think since output sharding prop is conditional, I would prefer keeping the non SPMD code path intact if possible.

**Comment by yeounoh** [comment]
The CI is failing, which seems real. I will address & add more unit test cases:
```
======================================================================
ERROR: test_mark_sharding_2d (__main__.BasicShardingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pytorch/xla/test/spmd/test_xla_sharding.py", line 60, in test_mark_sharding_2d
    actual = (xt1 + xt2).cpu()
RuntimeError: std::bad_cast

======================================================================
ERROR: test_mark_sharding_4d (__main__.BasicShardingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pytorch/xla/test/spmd/test_xla_sharding.py", line 78, in test_mark_sharding_4d
    actual = (xt + xt).cpu()
RuntimeError: std::bad_cast

======================================================================
ERROR: test_optimizer_step_with_sharding (__main__.BasicShardingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/pytorch/xla/test/spmd/test_xla_sharding.py", line 138, in test_optimizer_step_with_sharding
    xm.mark_step()
  File "/opt/conda/lib/python3.8/site-packages/torch_xla-2.1.0-py3.8-linux-x86_64.egg/torch_xla/core/xla_model.py", line 832, in mark_step
    torch_xla._XLAC._xla_step_marker(
RuntimeError: std::bad_cast
```

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:403]:
```
// Output parameter sharding annotations, defaults to REPLICATED(0) if unset.
auto tuple_shardings = ...
...
if (tuple_shardings[i].type()) {
      // Tensor sharding annotation type is non-zero (sharded).
```

File: torch_xla/csrc/ops/device_data.h
Comment (JackCaoG) [inline torch_xla/csrc/ops/device_data.h:28]: nit, let's also checkout about `!data_->hasValue()`, we should only call Assign on placeholders.

Comment (JackCaoG) [reply torch_xla/csrc/ops/device_data.h:28]: I take this as a short term solution. In the long term we should figure out why DeviceData is being generated by on a placeholder during a `SyncTensorGraphsInternal` call.

Comment (yeounoh) [reply torch_xla/csrc/ops/device_data.h:28]: Ah, for the current use case it's true -- do we want to restrict the API like that? Alternatively, we can pass the ownership (move) if the data is not a placeholder if someone tries to use the API. Hmm I am split, and see that the restriction would be safer semantically. WDYT?

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (JackCaoG) [inline torch_xla/csrc/xla_sharding_util.cpp:441]: Do we still expect this part of the code being executed? There shouldn't be any IR being generated when there is a placeholder in place. I think we should error out here instead of silently patching the node.

File: torch_xla/csrc/ops/device_data.h
Comment (yeounoh) [reply torch_xla/csrc/ops/device_data.h:28]: Will restrict it so that this is only used for placholders until we find more use cases (justification) cc @JackCaoG

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (yeounoh) [reply torch_xla/csrc/xla_sharding_util.cpp:441]: This is needed since we don't reset IR when we set placeholders -- https://github.com/pytorch/xla/blob/dc057dcda60355ea38d5686e5baafe677b838564/torch_xla/csrc/xla_graph_executor.cpp#L916

Comment (JackCaoG) [reply torch_xla/csrc/xla_sharding_util.cpp:450]: I am still a bit confuse in 
https://github.com/pytorch/xla/blob/dc057dcda60355ea38d5686e5baafe677b838564/torch_xla/csrc/xla_graph_executor.cpp#L965-L967
we reset IRs for all of the tensors that will get a new `XlaData` after the execution and in https://github.com/pytorch/xla/blob/dc057dcda60355ea38d5686e5baafe677b838564/torch_xla/csrc/xla_graph_executor.cpp#L915-L923 we assign the placeholder to those XlaTensor(which doesn't have any IR). 

In logic here we are replacing the `unsharded placeholder` with the `sharded_data_placeholder`. I am not sure in what case there will be a `DeviceData` Node that is pointed to the `unsharded placeholder`

**Comment by alanwaketan** [comment]
cc @khatwanimohit

File: torch_xla/csrc/xla_sharding_util.cpp
Comment (JackCaoG) [reply torch_xla/csrc/xla_sharding_util.cpp:465]: discussed offline with yeounoh, we will land this pr as it is and I will pick this us and fiugre out what is going on so we don't need to block the benchmark.

File: test/run_tests.sh
Comment (yeounoh) [inline test/run_tests.sh:191]: @JackCaoG we need to revisit SPMD+Dynamo integration after output sharding. Let's take it outside the current scope, as it requires some more work & design.

**Comment by yeounoh** [comment]
GPT-2 benchmark and basic sharding tests are passing. Noticed that this is breaking the current SPMD+Dynamo integration, I will take that outside the current scope and handle separately. We need to enable Dynamo with output sharding anyway. cc @JackCaoG

**Pull request merged**
